### PR TITLE
Message events (attempt 4)

### PR DESCRIPTION
**Pull request by rowanseymour**

### CODE REVIEW

File: temba/flows/models.py [hunk 1]
```diff
@@ -1882,7 +1882,7 @@ def start_msg_flow_batch(self, batch_contact_ids, broadcasts, started_flows, sta
                                partial_recipients=partial_recipients, run_map=run_map)
 
                 # map all the messages we just created back to our contact
-                for msg in Msg.objects.filter(broadcast=broadcast, created_on=created_on).select_related('channel'):
+                for msg in Msg.objects.filter(broadcast=broadcast, created_on=created_on).select_related('channel', 'contact_urn'):
```
Comment (rowanseymour) [inline temba/flows/models.py]: avoids additional db hits when it comes time to convert these messages to events

File: temba/flows/models.py [hunk 2]
```diff
@@ -1901,15 +1901,16 @@ def start_msg_flow_batch(self, batch_contact_ids, broadcasts, started_flows, sta
...
         optimize_sending_action = len(broadcasts) > 0
 
         for run in runs:
             contact = run.contact
 
             # each contact maintains its own list of started flows
             started_flows_by_contact = list(started_flows)
-            run_msgs = message_map.get(contact.id, [])
+            run_msgs = [start_msg] if start_msg else []
+            run_msgs += message_map.get(contact.id, [])
```
Comment (rowanseymour) [inline temba/flows/models.py]: tweaked this stuff so that messages are added to runs in the right order

File: temba/flows/models.py [hunk 3]
```diff
@@ -2012,25 +2012,29 @@ def add_step(self, run, node, msgs=None, exit_uuid=None, is_start=False, previou
...
-        if len(path) > FlowRun.PATH_MAX_STEPS:
-            path = path[len(path) - FlowRun.PATH_MAX_STEPS:]
+        if len(run.path) > FlowRun.PATH_MAX_STEPS:
+            run.path = run.path[len(run.path) - FlowRun.PATH_MAX_STEPS:]
+
+        update_fields = ['path', 'current_node_uuid']
+
+        if msgs:
+            run.add_messages(msgs, step=step, do_save=False)
+            update_fields += ['message_ids', 'responded', 'events']
```
Comment (rowanseymour) [inline temba/flows/models.py]: if we're calling `add_messages` from `add_step` we can defer the write to the run row

File: temba/flows/tests.py [hunk 4]
```diff
@@ -8206,7 +8339,7 @@ def test_flow_batch_start(self):
         stopped.stop(self.admin)
 
         # start our flow, this will take two batches
-        with QueryTracker(assert_query_count=308, stack_count=10, skip_unique_queries=True):
+        with QueryTracker(assert_query_count=298, stack_count=10, skip_unique_queries=True):
```
Comment (rowanseymour) [inline temba/flows/tests.py]: this is thanks to the `do_save` optimization on `add_messages`

File: temba/flows/migrations/0148_flowrun_events.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+    dependencies = [
+        ('flows', '0147_new_engine_changes'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='flowrun',
+            name='events',
+            field=temba.utils.models.JSONAsTextField(default=list, help_text='The events recorded on this run in JSON format', null=True),
```
Comment (nicpottier) [inline temba/flows/migrations/0148_flowrun_events.py]: You don't even want to use a normal JSON field huh?

File: temba/flows/models.py [hunk 6]
```diff
@@ -2853,8 +2863,12 @@ def create_or_update_from_goflow(cls, session, contact, run_output, run_events,
             path.append(step)
         current_node_uuid = path[-1][FlowRun.PATH_NODE_UUID]
 
+        # for now we only store message events
+        events = [e for e in run_output['events'] if e['type'] in ('msg_received', 'msg_created')]
```
Comment (nicpottier) [inline temba/flows/models.py]: porque?

File: temba/flows/migrations/0148_flowrun_events.py (hunk 5 above)
Comment (rowanseymour) [reply temba/flows/migrations/0148_flowrun_events.py]: I thought we said it wasn't worth it for fields we'd never index, but I think I'd prefer using real JSON here even if it does come with a small performance penalty - feels more future proof

File: temba/flows/models.py (hunk 6 above)
Comment (rowanseymour) [reply temba/flows/models.py]: was thinking we're only using this for tests and simulator runs... so don't start saving all events just yet until we're a little more confident we have the events correct?

**Comment by rowanseymour** [comment]
@nicpottier API output for runs is an even simpler view than what we store - it's just `node` and `time`

**Pull request merged**